### PR TITLE
[WIP] [v6r21?] Replacing pyGSI with M2Crypto

### PR DESCRIPTION
**Pull request by miloszz**
BEGINRELEASENOTES
Replacing pyGSI with M2Crypto in X509 certificates handling. 

*Core
CHANGE: X509* classes implemented using M2Crypto

*FrameworkSystem
CHANGE: Changes required by updated Core

ENDRELEASENOTES

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -3,18 +3,21 @@ git+https://:@gitlab.cern.ch:8443/fts/fts-rest.git#egg=fts3
 git+https://github.com/DIRACGrid/pyGSI.git#egg=GSI
 
 # From pypi
+asn1
 certifi
 CMRESHandler>=1.0.0b4
 coverage>=4.0.3
 elasticsearch-dsl>=2.0.0
 funcsigs
+future
```
Comment (fstagni) [inline requirements.txt]: Why is this needed?

File: Core/Security/X509Certificate.py [hunk 2]
```diff
@@ -2,20 +2,52 @@
...
 from DIRAC.Core.Utilities import Time
 from DIRAC.Core.Utilities import DErrno
 from DIRAC.ConfigurationSystem.Client.Helpers import Registry
 
+VOMS_EXTENSION_OID = '1.3.6.1.4.1.8005.100.100.5'
+VOMS_FQANS_OID = '1.3.6.1.4.1.8005.100.100.4'
+VOMS_GENERIC_ATTRS_OID = '1.3.6.1.4.1.8005.100.100.11'
+DOMAIN_COMPONENT_OID = '0.9.2342.19200300.100.1.25'
+ORGANIZATIONAL_UNIT_NAME_OID = '2.5.4.11'
+COMMON_NAME_OID = '2.5.4.3'
```
Comment (fstagni) [inline Core/Security/X509Certificate.py]: Oh... shit! What's all above? Where does it come from? There's definitely the need for some documentation here!

File: Core/DISET/private/Transports/SSL/SocketInfo.py [hunk 3]
```diff
@@ -65,8 +65,8 @@ def gatherPeerCredentials( self ):
         identitySubject = peerChain.getIssuerCert()['Value'].getSubjectNameObject()[ 'Value' ]
     else:
       identitySubject = peerChain.getCertInChain( 0 )['Value'].getSubjectNameObject()[ 'Value' ]
-    credDict = { 'DN' : identitySubject.one_line(),
-                 'CN' : identitySubject.commonName,
+    credDict = { 'DN' : str(identitySubject),
+                 'CN' : str(identitySubject).split('CN')[1][1:],
```
Comment (fstagni) [inline Core/DISET/private/Transports/SSL/SocketInfo.py]: you can do only one conversion at line 67. But is needed after all?

File: .travis.yml [hunk 4]
```diff
@@ -1,4 +1,12 @@
 language: python
```
Comment (fstagni) [inline .travis.yml]: This file, but also others in this PR are given executable mode. Is it needed?

File: Core/Security/X509Certificate.py [hunk 5]
```diff
@@ -32,28 +64,31 @@ def loadFromFile( self, certLocation ):
     Return : S_OK / S_ERROR
     """
     try:
-      fd = file( certLocation )
-      pemData = fd.read()
-      fd.close()
+      with file( certLocation ) as fd:
```
Comment (fstagni) [inline Core/Security/X509Certificate.py]: I think using 
```
with open( certLocation, 'r') as fd:
```
is more understandable

File: Core/Security/X509Chain.py [hunk 6]
```diff
@@ -30,7 +36,13 @@ def __init__( self, certList = False, keyObj = False ):
     self.__hash = False
     if certList:
       self.__loadedChain = True
-      self.__certList = certList
+      self.__certList = []
+      for cert in certList:
+        if type( cert ) != type( M2Crypto.X509.X509 ):
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: isinstance?

File: Core/Security/X509Chain.py [hunk 7]
```diff
@@ -57,30 +69,38 @@ def loadChainFromFile( self, chainLocation ):
     Return : S_OK / S_ERROR
     """
     try:
-      fd = file( chainLocation )
-      pemData = fd.read()
-      fd.close()
+      with open( chainLocation ) as fd:
+        pemData = fd.read()
     except Exception as e:
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: `except IOError`

File: Core/Security/X509Chain.py [hunk 8]
```diff
@@ -146,34 +167,38 @@ def loadProxyFromString( self, pemData ):
...
-      ext = crypto.X509Extension( "proxyCertInfo", extval )
-      extList.append( ext )
-    return extList
+      if rfc:
+        # this OID defines proxy as RFC proxy
+        ext =  M2Crypto.X509.new_extension( 'proxyCertInfo', 'critical, language:1.3.6.1.5.5.7.21.1', critical = 1 )
+        extStack.push(ext)
+      elif rfcLimited:
+        # this OID defines proxy as RFC Limited proxy
+        ext = M2Crypto.X509.new_extension( 'proxyCertInfo', 'critical, language:1.3.6.1.4.1.3536.1.1.1.9', critical = 1 )
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: Do you have a link to doc?

File: Core/Security/X509Chain.py [hunk 9]
```diff
@@ -403,31 +454,39 @@ def __checkProxyDN( self, certStep, issuerStep ):
...
+      if ext[ 'OK' ]:
+        ext = ext[ 'Value' ]
+      else:
+        return 0
+      contraint = [ line.split( ":" )[1].strip() for line in ext.get_value().split( "\n" ) if line.split( ":" )[0] == "Path Length Constraint" ]
+      if len( contraint ) == 0:
+        return 0
+      if self.__isRFC == None:
+        self.__isRFC = True
+      if contraint[0] == "1.3.6.1.4.1.3536.1.1.1.9":
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: doc? :)

File: Core/Security/X509Request.py [hunk 10]
```diff
@@ -4,33 +4,36 @@
...
+  """
+  Class representing X509 Certificate Request
+  """
 
   def __init__( self, reqObj = None, pkeyObj = None ):
     self.__valid = False
     self.__reqObj = reqObj
     self.__pkeyObj = pkeyObj
-    if reqObj and pkeyObj:
+    if reqObj and pkeyObj:  # isn't it a bit too liberal?
```
Comment (fstagni) [inline Core/Security/X509Request.py]: what do you mean?

File: Core/Security/X509Request.py [hunk 11]
```diff
@@ -4,33 +4,36 @@
...
+    self.__reqObj = M2Crypto.X509.Request()
     self.__reqObj.set_pubkey( self.__pkeyObj )
     if limited:
-      self.__reqObj.get_subject().insert_entry( "CN", "limited proxy" )
+      self.__reqObj.get_subject().add_entry_by_txt( field = "CN", type = M2Crypto.ASN1.MBSTRING_ASC, entry =  "limited proxy", len=-1, loc=-1, set=0 )
     else:
-      self.__reqObj.get_subject().insert_entry( "CN", "proxy" )
-    self.__reqObj.sign( self.__pkeyObj, "SHA256" )
+      self.__reqObj.get_subject().add_entry_by_txt( field = "CN", type = M2Crypto.ASN1.MBSTRING_ASC, entry =  "proxy", len=-1, loc=-1, set=0 )
+    self.__reqObj.sign( self.__pkeyObj, "sha256" )
```
Comment (fstagni) [inline Core/Security/X509Request.py]: upper case (old) or lower case (new) are the same?

File: Core/DISET/private/Transports/SSL/SocketInfo.py [hunk 12]
```diff
@@ -65,8 +65,8 @@ def gatherPeerCredentials( self ):
         identitySubject = peerChain.getIssuerCert()['Value'].getSubjectNameObject()[ 'Value' ]
```
Comment (fstagni) [inline Core/DISET/private/Transports/SSL/SocketInfo.py]: Regarding this file: is it the only one that needs to be changed for moving to M2Crypto?

**Comment by fstagni** [comment]
In general, we should try out full compatibility, which means:
- creating a proxy with pyGSI, using it in an M2Crypto world
- creating a proxy with M2Crypto, using it in an pyGSI world
Since there are changes involving the very base networking, it should be understood if the client/server interaction would still work in a mixed world (client and server may not share the same version).

We need to write integration tests with all the above in mind.

File: requirements.txt (hunk 1 above)
Comment (miloszz) [reply requirements.txt]: asn1 module requires it for builtins. I'm not sure why it haven't installed it as its dependency, but I clearly remember having to install it. Or was it that Travis didn't install it? It was quite log ago, I don't remember the details, but I'm sure it's because asn1 library (I uninstalled future on my test machine to verify and asn1 crashed on importing builtins).

File: .travis.yml (hunk 4 above)
Comment (miloszz) [reply .travis.yml]: My bad, I'll fix it.

File: Core/DISET/private/Transports/SSL/SocketInfo.py (hunk 3 above)
Comment (miloszz) [reply Core/DISET/private/Transports/SSL/SocketInfo.py]: Changed to one conversion. It is needed (at least for now) because other parts of code expect it to be a string, not a X509Name object. But this is a good time to redesign and use string only when printing, etc. Added to my to-do list.

File: Core/DISET/private/Transports/SSL/SocketInfo.py (hunk 12 above)
Comment (miloszz) [reply Core/DISET/private/Transports/SSL/SocketInfo.py]: Unfortunately no. This is only to fix a crash.

File: Core/Security/X509Chain.py (hunk 6 above)
Comment (miloszz) [reply Core/Security/X509Chain.py]: Fixed.

File: Core/Security/X509Certificate.py (hunk 5 above)
Comment (miloszz) [reply Core/Security/X509Certificate.py]: Done

File: Core/Security/X509Chain.py (hunk 7 above)
Comment (miloszz) [reply Core/Security/X509Chain.py]: Fixed

File: Core/Security/X509Chain.py [hunk 13]
```diff
@@ -30,7 +36,13 @@ def __init__( self, certList = False, keyObj = False ):
     self.__hash = False
     if certList:
       self.__loadedChain = True
-      self.__certList = certList
+      self.__certList = []
+      for cert in certList:
+        if isinstance( cert, M2Crypto.X509.X509 ):
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: if NOT isiinstance

File: Core/Security/X509Request.py (hunk 11 above)
Comment (miloszz) [reply Core/Security/X509Request.py]: M2Crypto doesn't recognise upper case name.

File: Core/Security/X509Chain.py (hunk 13 above)
Comment (miloszz) [reply Core/Security/X509Chain.py]: image

**Comment by fstagni** [comment]
Now the test is failing... ?

**Comment by miloszz** [comment]
remote: GitLab is not responding
fatal: unable to access 'https://:@gitlab.cern.ch:8443/fts/fts-rest.git/': The requested URL returned error: 502

**Comment by miloszz** [comment]
Sorry for review request, it made itself when I clicked "Update branch"...

**Comment by fstagni** [comment]
Please merge again integration: it will trigger more tests from Travis.

File: Core/Security/m2crypto/X509Certificate.py [hunk 14]
```diff
@@ -0,0 +1,539 @@
+""" X509Certificate is a class for managing X509 certificates alone
+"""
+__RCSID__ = "$Id$"
+
+import M2Crypto
+import asn1
+import datetime
+
+import os
```
Comment (fstagni) [inline Core/Security/m2crypto/X509Certificate.py]: Really minor: change the import orders (`os` and standard should go first)

File: Core/DISET/private/Transports/SSL/SocketInfo.py [hunk 15]
```diff
@@ -4,202 +4,218 @@
 import time
 import copy
 import os.path
+import os
```
Comment (fstagni) [inline Core/DISET/private/Transports/SSL/SocketInfo.py]: no need to import `os.path` if you import `os` (but then careful in the rest of rest of the module)

File: Core/Security/X509Chain.py [hunk 16]
```diff
@@ -125,92 +125,92 @@ def setPKey( self, pkeyObj ):
...
-    if diracGroup and isinstance( diracGroup, self.__validExtensionValueTypes ):
-      extList.append( crypto.X509Extension( 'diracGroup', diracGroup ) )
-    if rfc or rfcLimited:
-      blob = [ [ "1.3.6.1.5.5.7.21.1" ] ] if not rfcLimited else [ [ "1.3.6.1.4.1.3536.1.1.1.9" ] ]
-      asn1Obj = crypto.ASN1( blob )
+    extList.append(crypto.X509Extension('keyUsage',
+                                        'critical, digitalSignature, keyEncipherment, dataEncipherment'))
+    if diracGroup and isinstance(diracGroup, self.__validExtensionValueTypes):
+      extList.append(crypto.X509Extension('diracGroup', diracGroup))
+    if limited:
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: What do `limited` means here? Is it different from `rfcLimited` that was before?

File: Core/Security/X509Chain.py [hunk 17]
```diff
@@ -125,92 +125,92 @@ def setPKey( self, pkeyObj ):
...
+      blob = [["1.3.6.1.4.1.3536.1.1.1.9"]]
+      asn1Obj = crypto.ASN1(blob)
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: no need for variable `blob`, IIUC

File: Core/Security/X509Chain.py [hunk 18]
```diff
@@ -125,92 +125,92 @@ def setPKey( self, pkeyObj ):
...
-      extList.append( crypto.X509Extension( 'diracGroup', diracGroup ) )
-    if rfc or rfcLimited:
-      blob = [ [ "1.3.6.1.5.5.7.21.1" ] ] if not rfcLimited else [ [ "1.3.6.1.4.1.3536.1.1.1.9" ] ]
-      asn1Obj = crypto.ASN1( blob )
+    extList.append(crypto.X509Extension('keyUsage',
+                                        'critical, digitalSignature, keyEncipherment, dataEncipherment'))
+    if diracGroup and isinstance(diracGroup, self.__validExtensionValueTypes):
+      extList.append(crypto.X509Extension('diracGroup', diracGroup))
+    if limited:
+      blob = [["1.3.6.1.4.1.3536.1.1.1.9"]]
```
Comment (fstagni) [inline Core/Security/X509Chain.py]: Can you explain what this means?

File: Core/Security/m2crypto/X509Request.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
+""" X509Request is a class for managing X509 requests with their Pkeys
+"""
+
+__RCSID__ = "$Id$"
+
+import GSI
```
Comment (fstagni) [inline Core/Security/m2crypto/X509Request.py]: drop?

File: Core/Security/VOMS.py [hunk 20]
```diff
@@ -113,200 +116,193 @@ def getVOMSProxyInfo( self, proxy, option = False ):
...
+      cmdArgs.append('-voms "%s"' % vo)
+    cmdArgs.append('-valid "%s:%s"' % (hours, mins))
     tmpDir = False
     vomsesPath = self.getVOMSESLocation()
     if vomsesPath:
-      cmdArgs.append( '-vomses "%s"' % vomsesPath )
-    if chain.isRFC().get( 'Value' ):
-      cmdArgs.append( "-r" )
+      cmdArgs.append('-vomses "%s"' % vomsesPath)
+    cmdArgs.append("-r")  # XXX Not sure about that, maybe it's default
```
Comment (fstagni) [inline Core/Security/VOMS.py]: Keep it for now.

**Comment by fstagni** [comment]
There are conflicts.

File: Core/Security/X509Chain.py (hunk 16 above)
Comment (miloszz) [reply Core/Security/X509Chain.py]: No, the same thing, just the naming. rfcLimited was RFC and limited, to distinguish from non-RFC and limited. Now all are RFC, so that's simply limited.

File: Core/Security/X509Chain.py (hunk 18 above)
Comment (miloszz) [reply Core/Security/X509Chain.py]: It's limited proxy OID in SSL. I replaced it with defined value.

**Comment by miloszz** [comment]
I'm aware of the conflicts, but I'm in the middle of working and want to reach a committable state before rebasing/merging.
